query_executor.go

```diff
@@ -30,12 +34,28 @@ func (q *queryExecutor) attemptQuery(qry ExecutableQuery, conn *Conn) *Iter {
 }
 
 func (q *queryExecutor) executeQuery(qry ExecutableQuery) (*Iter, error) {
-	rt := qry.retryPolicy()
-	hostIter := q.policy.Pick(qry)
 
-	var iter *Iter
-	for hostResponse := hostIter(); hostResponse != nil; hostResponse = hostIter() {
-		host := hostResponse.Info()
+	// check whether the query execution set to speculative
+	// to make sure, we check if the policy is NonSpeculative or marked as non-idempotent, if it is
+	// run normal, otherwise, runs speculative.
+	_, nonSpeculativeExecution := qry.speculativeExecutionPolicy().(NonSpeculativeExecution)
+	if nonSpeculativeExecution || !qry.IsIdempotent() {
+		return q.executeNormalQuery(qry)
+	} else {
+		return q.executeSpeculativeQuery(qry)
+	}
+}
+
+func (q *queryExecutor) executeNormalQuery(qry ExecutableQuery) (*Iter, error) {
```

Comment:
I'm not keen on how this requires having 2 implementations of a very similar function to execute a query if it is speculatable.

Reply:
I totally agree. Just haven't found an elegant way to do it yet.